plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerTypeMapping.java

```diff
@@ -804,6 +804,63 @@ public void testSqlServerDatetimeOffset(ZoneId sessionZone)
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.999 UTC'", "TIMESTAMP '1970-01-01 00:00:00.999 UTC'")
+                // max supported precision
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123456 UTC'", "TIMESTAMP '1970-01-01 00:00:00.123456 UTC'")
+
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.1 UTC'", "TIMESTAMP '2020-09-27 12:34:56.1 UTC'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.9 UTC'", "TIMESTAMP '2020-09-27 12:34:56.9 UTC'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123 UTC'", "TIMESTAMP '2020-09-27 12:34:56.123 UTC'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123000 UTC'", "TIMESTAMP '2020-09-27 12:34:56.123000 UTC'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.999 UTC'", "TIMESTAMP '2020-09-27 12:34:56.999 UTC'")
+                // max supported precision
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.1234567 UTC'", "TIMESTAMP '2020-09-27 12:34:56.1234567 UTC'")
+
+                // round down
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.12345671 UTC'", "TIMESTAMP '1970-01-01 00:00:00.1234567 UTC'")
+
+                // nanoc round up, end result rounds down
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.1234567499 UTC'", "TIMESTAMP '1970-01-01 00:00:00.1234567 UTC'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123456749999 UTC'", "TIMESTAMP '1970-01-01 00:00:00.1234567 UTC'")
+
+                // round up
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.12345675 UTC'", "TIMESTAMP '1970-01-01 00:00:00.1234568 UTC'")
+
+                // max precision
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.111222333444 UTC'", "TIMESTAMP '1970-01-01 00:00:00.1112223 UTC'")
+
+                // round up to next second
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.99999995 UTC'", "TIMESTAMP '1970-01-01 00:00:01.0000000 UTC'")
+
+                // round up to next day
+                .addRoundTrip("TIMESTAMP '1970-01-01 23:59:59.99999995 UTC'", "TIMESTAMP '1970-01-02 00:00:00.0000000 UTC'")
+
+                // negative epoch
+                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.99999995 UTC'", "TIMESTAMP '1970-01-01 00:00:00.0000000 UTC'")
+                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.999999949999 UTC'", "TIMESTAMP '1969-12-31 23:59:59.9999999 UTC'")
+                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.99999994 UTC'", "TIMESTAMP '1969-12-31 23:59:59.9999999 UTC'")
+
+                // CTAS with Trino, where the coercion is done by the connector
+                .execute(getQueryRunner(), session, trinoCreateAsSelect("test_timestamp_tz"))
+                // INSERT with Trino, where the coercion is done by the engine
+                .execute(getQueryRunner(), session, trinoCreateAndInsert("test_timestamp_tz"));
```

Comment:
It seems the existing methods have a correctness issue for old timestamps. Could you please fix this issue first? 
```
trino:dbo> create table test (c1 timestamp with time zone);
CREATE TABLE
trino:dbo> insert into test values (TIMESTAMP '1400-09-27 12:34:56.123 UTC');
INSERT: 1 row

1> select * from database_d48540c5b41e42f788cb96c6eb937067.dbo.test;
2> go
c1
---------------------------------------------
               1400-09-18 12:34:56.123 +00:00
```

Reply:
Strange, I cannot reproduce it locally : 
```
trino:tpch> use sqlserver.dbo;
USE
trino:dbo> create table test (c1 timestamp with time zone);
CREATE TABLE
trino:dbo> insert into test values (TIMESTAMP '1400-09-27 12:34:56.123 UTC');
INSERT: 1 row

Query 20230315_100301_00030_a3ntz, FINISHED, 3 nodes
Splits: 12 total, 12 done (100.00%)
0.48 [0 rows, 0B] [0 rows/s, 0B/s]

trino:dbo> select * from test;
             c1
-----------------------------
 1400-09-27 12:34:56.123 UTC
(1 row)
```

But I see you are connected not to local instance (database_d48540c5b41e42f788cb96c6eb937067), maybe it occurs on different SqlServer versions?

Reply:
Please select on SQL Server, not Trino.

Reply:
> It seems the existing methods have a correctness issue for old timestamps #16559. Could you please fix this issue first?

Could it be done in parallel ? 
Because looks like this issue we already have and with current PR we didn't introduce it.
User can add values directly in SqlServer, but trino will fetch incorrect values.

Reply:
Do you mean fixing the issue in this PR? That's also fine.